### PR TITLE
Linkshadow

### PR DESCRIPTION
**Pull request by Devika-LS**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
A few sentences describing the overall goals of the pull request's commits.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 5.5.0
- [ ] 6.0.0
- [x] 6.1.0
- [ ] 6.2.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by mchasepan** [comment]
@marcellarichmond

File: Packs/Linkshadow/Integrations/Linkshadow/Linkshadow.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+        apiKey=apiKey,
+        api_username=username,
+        plugin_id=plugin_id,
+        action=action,
+        time_frame=time_frame
+    )
+
+    return CommandResults(
+        outputs_prefix='Linkshadow.anomaly',
+        outputs_key_field='demisto.args()',
```
Comment (Bargenish) [inline Packs/Linkshadow/Integrations/Linkshadow/Linkshadow.py:155]: this is a bug.  Please read https://xsoar.pan.dev/docs/integrations/code-conventions#commandresults for more information.

File: Packs/Linkshadow/Integrations/Linkshadow/Linkshadow.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+    else:
+        last_fetch = int(last_fetch)
+    latest_created_time = int(last_fetch)
+    incidents = []
+    alerts = client.fetch_anomaly(apiKey=apiKey, start_time=first_fetch_time,
+                                  api_username=api_username, plugin_id=plugin_id, action=action, time_frame=time_frame)
+    alert = {}
+    for alert in alerts:
+
+        if alert == 'data':
```
Comment (Bargenish) [inline Packs/Linkshadow/Integrations/Linkshadow/Linkshadow.py:96]: This for loop is redundant.

File: Packs/Linkshadow/Integrations/Linkshadow/Linkshadow.yml [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+configuration:
+- display: API Key
+  name: apiKey
+  type: 4
+  required: true
+- display: URL
+  name: url
+  type: 0
+  required: true
+- display: ACTION
```
Comment (Bargenish) [inline Packs/Linkshadow/Integrations/Linkshadow/Linkshadow.yml:28]: If there's a list of possible actions, I recommend defining a list (type 15).
https://xsoar.pan.dev/docs/integrations/yaml-file

**Review by Bargenish — CHANGES_REQUESTED** [review]
@Devika-LS Thank you for your contribution.
* Please fill the contribution [registration form](https://docs.google.com/forms/d/e/1FAIpQLSd0V270r75tTy9YbT50irQBUpKK8Wfn1aCUihYNn2rc_LkPpA/viewform)
* The content you have in Packs/Linkshadow/README.md is more suitable to the integration's README file + you need to add the integration's README file. You can follow https://xsoar.pan.dev/docs/documentation/pack-docs and https://xsoar.pan.dev/docs/documentation/readme_file for examples and guidance.
* Packs/Linkshadow/Integrations/Linkshadow/Linkshadow_image.png - Please provide an image with a transparent background.


See my comments below.
Let me know if you need any assistance.

File: Packs/Linkshadow/Integrations/Linkshadow/Linkshadow.py (hunk 2 above)
Comment (Devika-LS) [reply Packs/Linkshadow/Integrations/Linkshadow/Linkshadow.py:95]: When we call Linkshadow API we will get a List of Anomalies as response . Then we are looping the response and create each anomaly as an Incident in XSOAR (also we are mapping Anomaly fields to XSOAR custom fields)

**Review by Devika-LS — COMMENTED** [review]
Updated upon the request made by Bargenish

File: Packs/Linkshadow/Integrations/Linkshadow/Linkshadow.py (hunk 2 above)
Comment (Bargenish) [reply Packs/Linkshadow/Integrations/Linkshadow/Linkshadow.py:95]: Yeah, I meant that only the first loop is redundant.
```suggestion
    for alert in alerts.get('data'):
```

File: Packs/Linkshadow/Integrations/Linkshadow/Linkshadow.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    alerts = client.fetch_anomaly(
+        apiKey=apiKey,
+        api_username=username,
+        plugin_id=plugin_id,
+        action=action,
+        time_frame=time_frame
+    )
+
+    results = CommandResults(
+        outputs_prefix='Linkshadow.anomaly',
```
Comment (Bargenish) [inline Packs/Linkshadow/Integrations/Linkshadow/Linkshadow.py:153]: This should and does not match the context path defined in the yml file - `Linkshadow.data.FIELD_NAME`.
let's change this (in the yml definition or here) to match the object that gets returned.

File: Packs/Linkshadow/Integrations/Linkshadow/Linkshadow.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+        outputs_key_field='demisto.args()',
```
Comment (Bargenish) [inline Packs/Linkshadow/Integrations/Linkshadow/Linkshadow.py:154]: This should be set as one of the anomaly identification fields. e.g `GlobalID`

**Review by Bargenish — CHANGES_REQUESTED** [review]
@Devika-LS Great job. see my comment below.
Let's schedule a short session.

**Comment by Bargenish** [comment]
@Devika-LS sent you a message on Slack to coordinate a time.

**Comment by Bargenish** [comment]
A demo was scheduled for Thursday at 10:00 AM Israel time.

File: Packs/Linkshadow/Integrations/Linkshadow/Linkshadow.yml [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+      in an API response. Use of this command will return the JSON structure of the
+      API response .
+  dockerimage: demisto/python3:3.9.5.20070
+  isfetch: true
+  runonce: false
+  script: '-'
+  subtype: python3
+  type: python
+  isFetchSamples: true
+fromversion: 5.5.0
```
Comment (Bargenish) [inline Packs/Linkshadow/Integrations/Linkshadow/Linkshadow.yml:171]:
```suggestion
fromversion: 5.5.0
tests:
- No test

```

**Review by Devika-LS — COMMENTED** [review]
updated!

File: Packs/Linkshadow/Integrations/Linkshadow/Linkshadow.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+
+    last_fetch = last_run.get('last_fetch', None)
+    if last_fetch is None:
+        last_fetch = first_fetch_time
+    else:
+        last_fetch = int(last_fetch)
+    latest_created_time = int(last_fetch)
+    incidents = []
+    alerts = client.fetch_anomaly(apiKey=apiKey, api_username=api_username,
+                                  plugin_id=plugin_id, action=action, time_frame=time_frame)
```
Comment (Bargenish) [inline Packs/Linkshadow/Integrations/Linkshadow/Linkshadow.py:90]: * What is the difference between the `time_frame` var and the `first_fetch_time`
* in the case where  `time_frame = 1 minute`, does it mean that I only pull anomalies from the last 1 minute?
  if so, the last run has no effect here.

File: Packs/Linkshadow/Integrations/Linkshadow/Linkshadow.py [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+    for dic in alerts['data']:
```
Comment (Bargenish) [inline Packs/Linkshadow/Integrations/Linkshadow/Linkshadow.py:91]:
```suggestion
    for dic in alerts.get('data'):
```

File: Packs/Linkshadow/Integrations/Linkshadow/Linkshadow.py [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+                        'destinationport': formatted_JSON['dport'],
+                        'macaddress': formatted_JSON['smac'],
+                        'alertid': formatted_JSON['anomaly_id'],
+                        'subcategory': formatted_JSON['category']
+                    }
+                }
+                incidents.append(incident)
+                # Update last run and add incident if the incident is newer than last fetch
+                # if incident_created_time > latest_created_time:
+                #     latest_created_time = incident_created_time
```
Comment (Bargenish) [inline Packs/Linkshadow/Integrations/Linkshadow/Linkshadow.py:119]: why is that commented? the `latest_created_time` is not being updated

File: Packs/Linkshadow/Integrations/Linkshadow/Linkshadow.py [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+        action=action,
+        time_frame=time_frame
+    )
+
+    alert = []
+    for dic in alerts.get('data'):
+        alert.append(dic)
+
+    if len(alert) == 0:
+        alert = [{"message": "Linkshadow Anomaly already acknowledged!!"}]
```
Comment (Bargenish) [inline Packs/Linkshadow/Integrations/Linkshadow/Linkshadow.py:152]: Unnecessary. the outcome of the alert list will be the same as `alerts.get('data')`
```suggestion
    
```

File: Packs/Linkshadow/Integrations/Linkshadow/Linkshadow.py [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+
+    return CommandResults(
+        outputs_prefix='Linkshadow.data',
+        outputs_key_field='GlobalID',
+        outputs=alert
```
Comment (Bargenish) [inline Packs/Linkshadow/Integrations/Linkshadow/Linkshadow.py:157]:
```suggestion
        outputs=alerts.get('data') or [{"message": "Linkshadow Anomaly already acknowledged!!"}]
```

**Review by Bargenish — APPROVED** [review]

**Pull request merged**